### PR TITLE
task: Add banner encouraging edge upgrade

### PR DESCRIPTION
**Pull request by chriswk**
Only triggers if there is any rows in client instances that have

    sdk_version: unleash-edge with version < 17.0.0

The function that checks this memoizes the check for 10 minutes to avoid scanning the client instances table too often.

### CODE REVIEW

**Comment by chriswk** [comment]
Don't approve yet, I'm fixing the possible SQL Injection first.

File: src/lib/features/metrics/instance/instance-service.ts [hunk 1]
```diff
@@ -224,4 +225,20 @@ export default class ClientInstanceService {
...
+        const semver = new SemVer(sdkVersion);
+        const instancesOfSdk =
+            await this.clientInstanceStore.getBySdkName(sdkName);
+        return instancesOfSdk.some((instance) => {
+            if (instance.sdkVersion) {
+                const [_sdkName, sdkVersion] = instance.sdkVersion.split(':');
+                const instanceUsedSemver = new SemVer(sdkVersion);
+                return instanceUsedSemver < semver;
+            }
+        });
```
Comment (gastonfournier) [inline src/lib/features/metrics/instance/instance-service.ts:242]: I was just checking that SemVer can throw some exceptions: https://github.com/npm/node-semver/blob/main/classes/semver.js and we have a helper class that does the try catch: https://github.com/Unleash/unleash/blob/chore/update-features-created-by-user-id/src/lib/util/semver.ts maybe we should just try catch here

File: frontend/src/component/banners/internalBanners/InternalBanners.tsx [hunk 2]
```diff
@@ -11,6 +22,10 @@ export const InternalBanners = () => {
                 .map((banner) => (
                     <Banner key={banner.id} banner={banner} />
                 ))}
+            <ConditionallyRender
+                condition={displayUpgradeEdgeBanner}
+                show={<Banner key={'upgradeEdge'} banner={upgradeEdgeBanner} />}
+            />
```
Comment (nunogois) [inline frontend/src/component/banners/internalBanners/InternalBanners.tsx:28]: Similarly to https://github.com/Unleash/unleash/pull/5239#discussion_r1413730809 - I would recommend we create a new `EdgeUpgradeBanner` (or similar) component and add it directly to `App.tsx`. I don't think it makes much sense to add it here, especially as it is not an [internal banner](https://docs.getunleash.io/reference/banners).

**Review by gastonfournier — APPROVED** [review]
Looks good just a small thing around handling SemVer I think it's a safety check that's not hurtful to add. But Granting the approval as I don't expect that to happen, but just being extra cautious could be a good thing here

File: frontend/src/interfaces/uiConfig.ts [hunk 3]
```diff
@@ -79,6 +79,7 @@ export type UiFlags = {
     executiveDashboard?: boolean;
     changeRequestConflictHandling?: boolean;
     feedbackComments?: Variant;
+    displayUpgradeEdgeBanner?: boolean;
```
Comment (nunogois) [inline frontend/src/interfaces/uiConfig.ts:82]: Might be worth considering making this a `Variant` flag instead. That way we can control not only whether the banner is shown, but also its contents through the flag.

**Review by nunogois — CHANGES_REQUESTED** [review]
LG, but like mentioned in https://github.com/Unleash/unleash/pull/6018/files#r1464683982, I don't think it belongs inside `InternalBanners`.

File: src/lib/features/metrics/instance/instance-service.ts (hunk 1 above)
Comment (chriswk) [reply src/lib/features/metrics/instance/instance-service.ts:242]: Thanks. Good point. I'll have a look

File: frontend/src/component/banners/internalBanners/InternalBanners.tsx (hunk 2 above)
Comment (chriswk) [reply frontend/src/component/banners/internalBanners/InternalBanners.tsx:28]: Agreed. Will fix

File: frontend/src/interfaces/uiConfig.ts (hunk 3 above)
Comment (chriswk) [reply frontend/src/interfaces/uiConfig.ts:82]: This is something we'd like to have in OSS deploys as well, and hopefully not for very long.

Comment (chriswk) [reply frontend/src/interfaces/uiConfig.ts:82]: So, I think having it as a normal flag is fine.

**Comment by chriswk** [comment]
> LG, but like mentioned in https://github.com/Unleash/unleash/pull/6018/files#r1464683982, I don't think it belongs inside `InternalBanners`.

It's no longer in InternalBanners.

File: frontend/src/component/banners/EdgeUpgradeBanner/EdgeUpgradeBanner.tsx [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+        message: `We noticed that you're using an outdated Unleash Edge. To ensure you continue to receive metrics, we recommend upgrading to v17.0.0 or later.`,
+        link: 'https://github.com/Unleash/unleash-edge',
+        linkText: 'Get latest',
+        variant: 'warning',
+    };
+    return (
+        <>
+            <ConditionallyRender
+                condition={displayUpgradeEdgeBanner}
+                show={<Banner key={'upgradeEdge'} banner={upgradeEdgeBanner} />}
```
Comment (nunogois) [inline frontend/src/component/banners/EdgeUpgradeBanner/EdgeUpgradeBanner.tsx:18]: No need for key here, since it's only 1 element.

**Review by nunogois — APPROVED** [review]
LGTM! Thanks for addressing my comment 🙌

**Pull request merged**